scale-decode/src/visitor/decode.rs

```diff
@@ -168,6 +202,20 @@ fn decode_primitive_value<'scale, 'info, V: Visitor>(
         Ok(arr)
     }
 
+    // guard against invalid composite types: only `U8`, `U16`, `U32`, `U64` and `U128` can be compact encoded
+    if is_compact
+        && !matches!(
+            ty,
+            TypeDefPrimitive::U8
+                | TypeDefPrimitive::U16
+                | TypeDefPrimitive::U32
+                | TypeDefPrimitive::U64
+                | TypeDefPrimitive::U128
+        )
+    {
+        return Err(DecodeError::CannotDecodeCompactIntoType(ty.clone().into()).into());
+    }
```

Comment:
Personally for this and the `matches!` above, I'd be tempted to do something like define this here:

```rust
macro_rules err_if_compact! {
    () => {
        if is_compact {
            return Err(DecodeError::CannotDecodeCompactIntoType(ty.clone().into()).into());
        }
    }
}
```

And then in each branch wehre it applies:

```rust
match ty {
    // ...
    TypeDefPrimitive::Bool => {
        err_if_compact!();
        // usual logic here
    }
    // ...
}
```

The reason for this is so that it's very clear that all of the branches that don't want/have any special compact logic are guarded right there, rather than having the check a little bit away from where it's protecting :)

(also it saves us a match but that's not a big deal)